src/common.cpp

```diff
@@ -114,20 +158,63 @@ instruction_ref insert_common_op(module& m,
                                  const operation& op,
                                  std::vector<instruction_ref> inputs)
 {
-    auto common = common_shape(to_shapes(inputs));
-    std::transform(inputs.begin(), inputs.end(), inputs.begin(), [&](auto input) {
-        if(input->get_shape().lens() != common.lens())
+    if(std::any_of(
+           inputs.cbegin(), inputs.cend(), [](auto input) { return input->get_shape().dynamic(); }))
+    {
+        // currently only handles the binary case
+        if(inputs.size() != 2)
         {
-            input = m.insert_instruction(
-                ins, make_op("multibroadcast", {{"out_lens", common.lens()}}), input);
+            MIGRAPHX_THROW("INSERT_COMMON_OP: not handled; " + migraphx::to_string(inputs.size()) +
+                           "inputs, only handle two inputs if any are dynamic shape");
         }
-        if(input->get_shape().type() != common.type())
+
+        auto c_type = compute_common_types(to_shapes(inputs));
+        auto c_dyn_dims =
+            compute_broadcasted_dyn_dims(inputs[0]->get_shape(), inputs[1]->get_shape());
+
+        // following should work for a static or dynamic shape
+        if(inputs[0]->get_shape().dyn_dims() != c_dyn_dims)
         {
-            input = m.insert_instruction(
-                ins, make_op("convert", {{"target_type", common.type()}}), input);
+            inputs[0] = m.insert_instruction(
+                ins,
+                make_op("multibroadcast", {{"out_dyn_dims", to_value(c_dyn_dims)}}),
+                inputs[0],
+                inputs[1]);
         }
-        return input;
-    });
+        if(inputs[1]->get_shape().dyn_dims() != c_dyn_dims)
+        {
+            inputs[1] = m.insert_instruction(
+                ins,
+                make_op("multibroadcast", {{"out_dyn_dims", to_value(c_dyn_dims)}}),
+                inputs[1],
+                inputs[0]);
```

Comment:
These two `IFs` can be combined into single one. Just check if either of the inputs is not the same as `c_dyn_dims`. 

`multibroadcast` would broadcast both of them to same common shape.

Reply:
So there are two `if` statements right now to avoid having two `multibroadcast` calls if one of inputs is already correct. This is to mimic the behavior for the static shape stuff. The two input version of `multibroadcast` only reshapes the first input because it can only have one output.

Reply:
Okay, but then we shouldn't be making use of  two input version call i guess.It is confusing.

Reply:
It has to because it needs to broadcast to the other shape at model evaluation time.
example:
shape a is dynamic [{1, 4, 0}, {4, 4, 0}, {4, 4, 0}]
shape b is fixed [4, 4]

* multibroadcast(a -> b) is not required, because it would do nothing
* only multibroadcast(b -> a) is required
* however b could broadcast to any shape in dynamic range of a: [1, 4, 4] to [4, 4, 4]